src/groupeddataframe/grouping.jl

```diff
@@ -387,6 +423,98 @@ function gen_fun(::Type{NT}, f2) where NT
     end
 end
 
+# Generate a named tuple of reductions to call for each group
+function gen_fun(::Type{NT}, f2::Tuple{Vararg{Pair{<:Val, <:Reduce}}}) where NT
+    tup = map(f2) do p
+        nms = get_val(first(p))
+        nms => last(p)
+    end
+    NT(tup)
+end
+
+# Use reducedim_init to get a vector of the right type,
+# but trick it into using the expected length
+groupreduce_init(op, incol, gd) =
+    Base.reducedim_init(identity, op, view(incol, 1:length(gd)), 2)
+
+for (op, initf) in ((:max, :typemin), (:min, :typemax))
+    @eval begin
+        function groupreduce_init(::typeof($op), incol::AbstractVector{T}, gd) where T
+            outcol = similar(incol, length(gd))
+            # Comparison is possible only between CatValues from the same pool
+            if incol isa CategoricalVector
+                U = Union{CategoricalArrays.leveltype(outcol),
+                          eltype(outcol) >: Missing ? Missing : Union{}}
+                outcol = CategoricalArray{U, 1}(outcol.refs, incol.pool)
+            end
+            # It is safe to use a non-missing init value
+            # since missing will poison the result if present
+            S = Missings.T(T)
+            if isconcretetype(S) && hasmethod($initf, Tuple{S})
+                return fill!(outcol, $initf(S))
+            else
+                @inbounds for i in eachindex(outcol)
+                    outcol[i] = incol[gd.idx[gd.starts[i]]]
+                end
+                return outcol
+            end
+        end
+    end
+end
+
+function copyto_widen!(res::AbstractVector{T}, x::AbstractVector) where T
+    @inbounds for i in eachindex(res, x)
+        val = x[i]
+        S = typeof(val)
+        if S <: T || promote_type(S, T) <: T
+            res[i] = val
+        else
+            newres = Tables.allocatecolumn(promote_type(S, T), length(x))
+            return copyto_widen!(newres, x)
+        end
+    end
+    return res
+end
+
+function groupreduce(op, adjust, incol::AbstractVector{T}, gd::GroupedDataFrame) where T
+    n = length(gd)
+    outcol = groupreduce_init(op, incol, gd)
+    if adjust !== nothing
+        counts = zeros(Int, n)
+    end
+    @inbounds @simd for i in eachindex(incol, gd.groups)
+        gix = gd.groups[i]
+        adjust !== nothing && (counts[gix] += 1)
+        outcol[gix] = op(outcol[gix], incol[i])
+    end
+    if adjust !== nothing
+        map!(adjust, outcol, outcol, counts)
+    end
+    # Undo pool sharing done by groupreduce_init
+    if outcol isa CategoricalVector
+        U = Union{CategoricalArrays.leveltype(outcol),
+                  eltype(outcol) >: Missing ? Missing : Union{}}
+        outcol = CategoricalArray{U, 1}(outcol.refs, incol.pool)
+    end
+    if isconcretetype(T)
+        return outcol
+    else
+        copyto_widen!(Tables.allocatecolumn(typeof(first(outcol)), n), outcol)
```

Comment:
This is also not optimal, but it's required if we want to return the same element type as the non-optimized method. Hopefully it should be quite rare, since reducing over `Union{T,Missing}` columns will return `missing` if there is any missing value, which isn't very useful (though one can have a  `Union{T,Missing}` column without any missing value).